File: sdk/typescript/src/phone/resources/numbers.ts
/**
 * inkbox-phone/resources/numbers.ts
 *
 * Phone number CRUD, provisioning, release, and transcript search.
 */

import { HttpTransport } from "../../_http.js";
import {
  PhoneNumber,
  PhoneTranscript,
  RawPhoneNumber,
  RawPhoneTranscript,
  parsePhoneNumber,
  parsePhoneTranscript,
} from "../types.js";

const BASE = "/numbers";

export class PhoneNumbersResource {
  constructor(private readonly http: HttpTransport) {}

  /** List all phone numbers for your organisation. */
  async list(): Promise<PhoneNumber[]> {
    const data = await this.http.get<RawPhoneNumber[]>(BASE);
    return data.map(parsePhoneNumber);
  }

  /** Get a phone number by ID. */
  async get(phoneNumberId: string): Promise<PhoneNumber> {
    const data = await this.http.get<RawPhoneNumber>(
      `${BASE}/${phoneNumberId}`,
    );
    return parsePhoneNumber(data);
  }

  /**
   * Update phone number settings. Only provided fields are updated.
   * Pass a field as `null` to clear it.
   *
   * @param phoneNumberId - UUID of the phone number.
   * @param options.incomingCallAction - `"auto_accept"`, `"auto_reject"`, or `"webhook"`.
   * @param options.clientWebsocketUrl - WebSocket URL (wss://) for audio bridging.
   * @param options.incomingCallWebhookUrl - Webhook URL called for incoming calls when action is `"webhook"`.
   */
  async update(
    phoneNumberId: string,
    options: {
      incomingCallAction?: string;
      clientWebsocketUrl?: string | null;
      incomingCallWebhookUrl?: string | null;
      incomingTextWebhookUrl?: string | null;
    },
  ): Promise<PhoneNumber> {
    const body: Record<string, unknown> = {};
    if (options.incomingCallAction !== undefined) {
      body["incoming_call_action"] = options.incomingCallAction;
    }
    if ("clientWebsocketUrl" in options) {
      body["client_websocket_url"] = options.clientWebsocketUrl;
    }
    if ("incomingCallWebhookUrl" in options) {
      body["incoming_call_webhook_url"] = options.incomingCallWebhookUrl;
    }
    if ("incomingTextWebhookUrl" in options) {
      body["incoming_text_webhook_url"] = options.incomingTextWebhookUrl;
    }
    const data = await this.http.patch<RawPhoneNumber>(
      `${BASE}/${phoneNumberId}`,
      body,
    );
    return parsePhoneNumber(data);
  }

  /**
   * Provision a new phone number and link it to an agent identity.
   *
   * @param options.agentHandle - Handle of the agent identity to assign this number to.
   * @param options.type - `"toll_free"` or `"local"`. Defaults to `"toll_free"`.
   * @param options.state - US state abbreviation (e.g. `"NY"`). Only valid for `local` numbers.
   */
  async provision(options: {
    agentHandle: string;
    type?: string;
    state?: string;
  }): Promise<PhoneNumber> {
    const body: Record<string, unknown> = {
      agent_handle: options.agentHandle,
      type: options.type ?? "toll_free",
    };
    if (options.state !== undefined) {
      body["state"] = options.state;
    }
    const data = await this.http.post<RawPhoneNumber>(BASE, body);
    return parsePhoneNumber(data);
  }

  /**
   * Release a phone number.
   *
   * @param phoneNumberId - UUID of the phone number to release.
   */
  async release(phoneNumberId: string): Promise<void> {
    await this.http.delete(`${BASE}/${phoneNumberId}`);
  }

  /**
   * Full-text search across transcripts for a phone number.
   *
   * @param phoneNumberId - UUID of the phone number.
   * @param options.q - Search query string.
   * @param options.party - Filter by speaker: `"local"` or `"remote"`.
   * @param options.limit - Maximum number of results (1–200). Defaults to 50.
   */
  async searchTranscripts(
    phoneNumberId: string,
    options: { q: string; party?: string; limit?: number },
  ): Promise<PhoneTranscript[]> {
    const data = await this.http.get<RawPhoneTranscript[]>(
      `${BASE}/${phoneNumberId}/search`,
      { q: options.q, party: options.party, limit: options.limit ?? 50 },
    );
    return data.map(parsePhoneTranscript);
  }
}

File: sdk/rust/src/phone/resources/numbers.rs
//! Phone number CRUD, provisioning, release, and transcript search.

use std::sync::Arc;

use serde_json::{Map, Value};

use crate::error::Result;
use crate::http::HttpTransport;
use crate::phone::types::{FilterMode, PhoneNumber, PhoneTranscript};

const BASE: &str = "/numbers";

pub struct PhoneNumbersResource {
    http: Arc<HttpTransport>,
}

impl PhoneNumbersResource {
    pub fn new(http: Arc<HttpTransport>) -> Self {
        Self { http }
    }

    /// List all phone numbers for your organisation.
    pub fn list(&self) -> Result<Vec<PhoneNumber>> {
        let data = self.http.get(BASE, crate::http::NO_QUERY)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Get a phone number by ID.
    pub fn get(&self, phone_number_id: &str) -> Result<PhoneNumber> {
        let data = self
            .http
            .get(&format!("{BASE}/{phone_number_id}"), crate::http::NO_QUERY)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Update phone number settings.
    ///
    /// Pass only the fields you want to change; omitted fields (outer `None`)
    /// are left as-is. For the nullable string fields, `Some(None)` sends an
    /// explicit JSON `null` to clear the field, while `Some(Some(v))` sets it —
    /// mirroring the Python `_UNSET` sentinel vs an explicit `None` value.
    ///
    /// # Arguments
    /// * `incoming_call_action` - `"auto_accept"`, `"auto_reject"`, or `"webhook"`.
    /// * `client_websocket_url` - WebSocket URL (wss://) for audio bridging.
    /// * `incoming_call_webhook_url` - Webhook URL called for incoming calls when
    ///   action is `"webhook"`.
    /// * `filter_mode` - `whitelist` or `blacklist`. Admin-only on the server;
    ///   agent-scoped keys receive 403. Governs both inbound voice and SMS.
    ///
    /// # Returns
    /// The updated phone number. When `filter_mode` was supplied and actually
    /// changed, `filter_mode_change_notice` is populated.
    pub fn update(
        &self,
        phone_number_id: &str,
        incoming_call_action: Option<Option<&str>>,
        client_websocket_url: Option<Option<&str>>,
        incoming_call_webhook_url: Option<Option<&str>>,
        filter_mode: Option<FilterMode>,
    ) -> Result<PhoneNumber> {
        let mut body = Map::new();
        // Each nullable-string field: outer None omits, Some(None) -> JSON null,
        // Some(Some(v)) -> the value.
        if let Some(v) = incoming_call_action {
            body.insert("incoming_call_action".into(), str_or_null(v));
        }
        if let Some(v) = client_websocket_url {
            body.insert("client_websocket_url".into(), str_or_null(v));
        }
        if let Some(v) = incoming_call_webhook_url {
            body.insert("incoming_call_webhook_url".into(), str_or_null(v));
        }
        if let Some(mode) = filter_mode {
            let mode = match mode {
                FilterMode::Whitelist => "whitelist",
                FilterMode::Blacklist => "blacklist",
            };
            body.insert("filter_mode".into(), mode.into());
        }
        let data = self.http.patch(&format!("{BASE}/{phone_number_id}"), &body)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Provision a new phone number and link it to an agent identity.
    ///
    /// # Arguments
    /// * `agent_handle` - Handle of the agent identity to assign this number to.
    /// * `type_` - Number type to provision. Only `"local"` is supported.
    /// * `state` - US state abbreviation (e.g. `"NY"`) to request a number in
    ///   that state.
    ///
    /// # Returns
    /// The provisioned phone number.
    pub fn provision(
        &self,
        agent_handle: &str,
        type_: &str,
        state: Option<&str>,
    ) -> Result<PhoneNumber> {
        let mut body = Map::new();
        body.insert("agent_handle".into(), agent_handle.into());
        body.insert("type".into(), type_.into());
        if let Some(s) = state {
            body.insert("state".into(), s.into());
        }
        let data = self.http.post(BASE, Some(&body), crate::http::NO_QUERY)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Release a phone number.
    pub fn release(&self, phone_number_id: &str) -> Result<()> {
        self.http.delete(&format!("{BASE}/{phone_number_id}"))
    }

    /// Full-text search across transcripts for a phone number.
    ///
    /// # Arguments
    /// * `q` - Search query string.
    /// * `party` - Filter by speaker: `"local"` or `"remote"`.
    /// * `limit` - Maximum number of results (1-200).
    pub fn search_transcripts(
        &self,
        phone_number_id: &str,
        q: &str,
        party: Option<&str>,
        limit: i64,
    ) -> Result<Vec<PhoneTranscript>> {
        // Python passes `party=None` through; the transport drops None-valued
        // params, so we only push `party` when present.
        let mut params: Vec<(&str, String)> =
            vec![("q", q.to_string()), ("limit", limit.to_string())];
        if let Some(p) = party {
            params.push(("party", p.to_string()));
        }
        let data = self
            .http
            .get(&format!("{BASE}/{phone_number_id}/search"), &params)?;
        Ok(serde_json::from_value(data)?)
    }
}

/// Map an `Option<&str>` to a JSON string or an explicit JSON `null`.
fn str_or_null(v: Option<&str>) -> Value {
    match v {
        Some(s) => Value::String(s.to_string()),
        None => Value::Null,
    }
}
